satsense/features/sift.py

```diff
@@ -5,33 +5,51 @@
 import numpy as np
 from sklearn.cluster import MiniBatchKMeans
 
+from ..generators import FullGenerator
 from ..image import Image
 from .feature import Feature
 
-SIFT = cv2.xfeatures2d.SIFT_create()
-
 
-def sift_cluster(images: Iterator[Image], n_clusters=32,
-                 sample_size=100000) -> MiniBatchKMeans:
+def sift_cluster(images: Iterator[Image],
+                 n_clusters=32,
+                 sample_size=100000,
+                 sample_window=(8192, 8192)) -> MiniBatchKMeans:
     """Create the clusters needed to compute the sift feature."""
-    descriptors = None
+    nfeatures = int(sample_size / len(images))
+    sift_object = cv2.xfeatures2d.SIFT_create(nfeatures)
+
+    descriptors = []
     for image in images:
-        array = image['gray_ubyte']
-        inverse_mask = (~array.mask).astype(np.uint8)
-        _, new_descriptors = SIFT.detectAndCompute(array, inverse_mask)
-        del _  # Free up memory
-
-        # Add descriptors if we already had some
-        if descriptors is None:
-            descriptors = new_descriptors
-        else:
-            descriptors = np.append(descriptors, new_descriptors, axis=0)
-
-    if descriptors.shape[0] > sample_size:
-        # Limit the number of descriptors to sample_size
-        # by randomly selecting some rows
-        descriptors = descriptors[np.random.choice(
-            descriptors.shape[0], sample_size, replace=False), :]
+        if image.shape[0] < sample_window[0]:
```

Comment:
Replace by a more consice
```python
chunk = np.minimum(sample_window, image.shape)
```
and use `chunk` where it says `sample_window` below, to prevent using the smaller sample window for ever subsequent image.